compiler.go

```diff
@@ -62,19 +62,41 @@ type Compiler struct {
 	// errors and ignores all warnings.
 	Reporter reporter.Reporter
 
-	// If true, source code information will be included in the resulting
-	// descriptors. Source code information is metadata in the file descriptor
-	// that provides position information (i.e. the line and column where file
-	// elements were defined) as well as comments.
+	// If unspecified or set to SourceInfoNone, source code information will not
+	// be included in the resulting descriptors. Source code information is
+	// metadata in the file descriptor that provides position information (i.e.
+	// the line and column where file elements were defined) as well as comments.
+	//
+	// If set to SourceInfoStandard, normal source code information will be
+	// included in the resulting descriptors. This matches the output of protoc
+	// (the reference compiler for Protocol Buffers). If set to
+	// SourceInfoMoreComments, the resulting descriptor will attempt to preserve
+	// as many comments as possible, for all elements in the file, not just for
+	// complete declarations.
 	//
 	// If Resolver returns descriptors or descriptor protos for a file, then
 	// those descriptors will not be modified. If they do not already include
 	// source code info, they will be left that way when the compile operation
 	// concludes. Similarly, if they already have source code info but this flag
 	// is false, existing info will be left in place.
-	IncludeSourceInfo bool
+	SourceInfoMode SourceInfoMode
 }
 
+// SourceInfoMode indicates how source code info is generated by a Compiler.
+type SourceInfoMode int
+
+const (
+	// SourceInfoNone indicates that no source code info is generated.
+	SourceInfoNone = SourceInfoMode(iota)
+	// SourceInfoStandard indicates that the standard source code info is
+	// generated, which includes comments only for complete declarations.
+	SourceInfoStandard
+	// SourceInfoExtraComments indicates that source code info is generated
+	// and will include comments for all elements (more comments than would
+	// be found in a descriptor produced by protoc).
+	SourceInfoExtraComments
```

Comment:
Should this reference 'detached' instead of 'extra' comments? Will we need to update this in light of the upstream change to protoc being accepted?

At some point I wonder if we're going to want to represent things like 'features' for protocompile where it deviates from upstream protoc - i.e. `protocompile.New(..., WithDetachedComments())`.

Reply:
I don't quite follow. `protoc` supports detached comments also. The feature here really is to capture _more_ comments. `protoc` only includes comments for "complete declarations". But this mode will include comments for every element possible.